Fixed texture clamping caused by texture coordinates being outside of (0,1) interval

Pasalc opened the pull request:
# Fix for issue #173. 
### Short description
Video gets clamped for **[2659503088](https://steamcommunity.com/sharedfiles/filedetails/?id=2659503088)** wallpaper. Problem is caused by texture coordinates being out of boundaries, resulting in clamp.
### Full description
Texture UV coordinates starts from **0** and ends at **1**. We will change _uend_ and see how it will affect video **[2667198601](https://steamcommunity.com/sharedfiles/filedetails/?id=2667198601)**.
If we have:
```
float ustart = 0.0f;
float uend = 1.0f;
float vstart = 1.0f;
float vend = 0.0f;
```
we will get full size video:
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/63db7eef-11aa-43ca-912d-8997f498e626)

And if we have:
```
float ustart = 0.0f;
float uend = 0.5f;
float vstart = 1.0f;
float vend = 0.0f;
```
we will get only left halve of video:
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/be2ae186-ee1e-4f41-962b-dee059a0519a)

And if we  try to render texture out of this boundries:
```
float ustart = 0.0f;
float uend = 1.5f;
float vstart = 1.0f;
float vend = 0.0f;
```
texture will be clamped to the edge (because _GL_CLAMP_TO_EDGE_ parametr in _CFBO.cpp_ is set):
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/d4129eaa-c145-459f-963b-e5ccdc1fa1ff)

### What caused problem
Previously, _ustart/uend_ or _vstart/vend_ could be less than **0** or bigger than **1** (for example _ustart_ could be **-0.16** and _uend_ could be **1.15**), resulting in texture clamping on _left/right_ or _top/bottom_ of window or screen(though I only had _top/bottom_ clamp). By making them equal to **0** or **1** accordingly, we make sure that **OpenGL** renders whole texture with _glDrawArrays_ and don't get out of boundaries so we have no clamping.

**Before:**
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/6e4e26db-1f3c-47a3-be8f-4a1d3fcb1f97)

**After:**
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/85314cdc-74f4-47f4-b396-b03fe7b1ff1c)

### Additional concerns
It probably shouldn't affect people from issue #81, but may be it's a good idea to ask if everything works fine for them.

Teddy-Kun commented:
I can confirm this fixes the pixel artifacts, however I don't think this approach is always the best as this stretches the wallpaper to fit the size given by either the monitor or window. I believe that there could/should be a command line flag to change the behavior. On OLED screens I imagine it could be desirable to just leave the pixels outside completely black, whereas for my current personal wallpaper I would prefer cropping the sides to not stretch out the image. But to not undermine your work, this is 100% a lot better than before and the code definitely looks much cleaner and simpler to me!

Pasalc commented:
Thank you for response! I would probably do as you suggested and add flags:
1. **fit the size**
2. **leave the pixels outside completely black**
3. if flag was not provided: **preserve things as they are now**

If i understand correctly, by
> cropping the sides to not stretch out the image

you meant flag _2_, but correct me if I got it wrong.

Pasalc reviewed (commented):
In [src/WallpaperEngine/Render/Drivers/Output/CGLFWWindowOutput.cpp](https://github.com/Almamu/linux-wallpaperengine/pull/188/commits/f6bff679ca86107300ad378299742aa09e5fc967#diff-95858db18b792a50254e3b15348d5348929f5a4374d44d605e63970b7f89d6f9) added:
`#include <GL/glew.h>`
because otherwise compilation fails with error:
```
In file included from /home/kermit/Projects/fun/linux-wallpaperengine/src/WallpaperEngine/Render/Drivers/CVideoDriver.h:3,
                 from /home/kermit/Projects/fun/linux-wallpaperengine/src/WallpaperEngine/Render/Drivers/Output/CGLFWWindowOutput.h:3,
                 from /home/kermit/Projects/fun/linux-wallpaperengine/src/WallpaperEngine/Render/Drivers/Output/CGLFWWindowOutput.cpp:2:
/usr/include/GL/glew.h:85:2: error: #error gl.h included before glew.h
   85 | #error gl.h included before glew.h
      |  ^~~~~
/usr/include/GL/glew.h:97:2: error: #error glext.h included before glew.h
   97 | #error glext.h included before glew.h
      |  ^~~~~

```

Pasalc commented:
Just realized that this issue might be related to #90 , so I would add another flag option and make **zoom fill** another option too. For now I just made it default. Looks fine to me, but I'm not sure it really works correctly, so more thought must be put into that.
It looks like this for different window sizes(I don't know their exact size):
![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/dbebde06-afaa-43e0-9eef-961b60778094)

![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/78b21d51-798c-4fc3-afbd-0a3108108f54)

![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/5c9a7f6d-a5c7-46f2-ae61-acd9f78266f0)

![image](https://github.com/Almamu/linux-wallpaperengine/assets/43461852/0b36b058-72b5-477e-889b-2f37e2ff55a8)

So yeah, it's either zoomed to full width or to full height.

Pasalc commented:
### Overview

So I added 2 flags:
```
--scaling <mode>	 Scaling mode for wallpaper. Can be stretch, fit, fill, default. Must be used before wallpaper provided.
                    		 Example: ./wallengine --scaling stretch --screen-root eDP-1 --bg 2667198601 --scaling fill --screen-root eDP-2 2667198602

--clamping <mode>	 Clamping mode for all wallpapers. Can be clamp, border, repeat. Enables GL_CLAMP_TO_EDGE, GL_CLAMP_TO_BORDER, GL_REPEAT accordingly. Default is clamp.
```

Scaling mode should work with **different** screens, so you could for example set _stretch_ for one wallpaper and _fill_ for other. You **must** provide `--scaling` option before wallpaper was provided. I haven't tested it for several screens now though. Also for screens which width is bigger than wallpaper width.

Clamping mode apply to **all** screens, so you can't set it for each screen individually.

### Code
Added new class CWallpaperState files:
- [CWallpaperState.cpp](https://github.com/Pasalc/linux-wallpaperengine/blob/viewport-clamp/src/WallpaperEngine/Render/CWallpaperState.cpp)
- [CWallpaperState.h](https://github.com/Pasalc/linux-wallpaperengine/blob/viewport-clamp/src/WallpaperEngine/Render/CWallpaperState.h)

Added **CWallpaperState** as new member in **CWallpaper** class. 

CWallpaperState saves last **viewport, projection, vflip** and **UVs** values. Now if _nothing_ have changed(_viewport_, _projection_, _vflip_) then old **UVs** coordinates are used (or more simple: **UVs** are cached). 
If something _have_ changed, then the value of **m_textureUVsMode** defines which scaling algorithm will be applied to recalculate **UVs**.

If any new scaling algorithm will be needed in the future, then it can be easily(I hope) added:

1.  Add new value to `enum class TextureUVsScaling` in [CWallpaperState.h](https://github.com/Pasalc/linux-wallpaperengine/blob/viewport-clamp/src/WallpaperEngine/Render/CWallpaperState.h)
2. Provide new template specialization for your algorithm with `template<> void CWallpaperState::updateTextureUVs<CWallpaperState::TextureUVsScaling::YOUR_ALGORITHM>`  in [CWallpaperState.cpp](https://github.com/Pasalc/linux-wallpaperengine/blob/viewport-clamp/src/WallpaperEngine/Render/CWallpaperState.cpp)
3. Add new _switch_ statement in `CWallpaperState::updateState(...)`
4. Add new _switch_ statement for `case 't'` in  [CApplicationContext.cpp](https://github.com/Pasalc/linux-wallpaperengine/blob/viewport-clamp/src/WallpaperEngine/Application/CApplicationContext.cpp)

Also added  `constexpr size_t customHash(const char* str)` in [CApplicationContext.cpp](https://github.com/Pasalc/linux-wallpaperengine/commit/f433218ce8e464417e1985e12ef982210ef78055#diff-7de3d2969c04d0ff7ce1a177348f7dab35f9afcb70dfbf9197149c230b90057a). Made it to have compile time hash for choosing options for _scaling_ and _clamping_ in switch statement. Could be done with simple _if else_ statements, so maybe it's a bit of an overkill.

Pasalc commented on src/WallpaperEngine/Render/Drivers/Output/CGLFWWindowOutput.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <GL/glew.h>
```
Won't compile without that

Pasalc commented on src/WallpaperEngine/Render/CWallpaperState.cpp:
```diff
@@ -0,0 +1,170 @@
+
+    const float m1 = static_cast<float>(viewportWidth) / projectionWidth;
+    const float m2 = static_cast<float>(viewportHeight) / projectionHeight;
+    const float m = std::max(m1,m2);
```
This (line 74) is the only difference with  `ZoomFitUVs`(line 98) specialization

Pasalc commented on src/WallpaperEngine/Render/CWallpaperState.cpp:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+template<> void CWallpaperState::updateTextureUVs<CWallpaperState::TextureUVsScaling::DefaultUVs>(){
```
This is copy of code that was used to calculate **UVs** before(in current main branch). Behavior(resulting **UVs** values) should be the same as old version

Pasalc commented on src/WallpaperEngine/Application/CApplicationContext.cpp:
```diff
@@ -290,4 +349,8 @@ void CApplicationContext::printHelp (const char* route)
     sLog.out ("\t--set-property <name=value>\tOverrides the default value of the given property");
     sLog.out ("\t--no-fullscreen-pause\tPrevents the background pausing when an app is fullscreen");
     sLog.out ("\t--disable-mouse\tDisables mouse interactions");
+    sLog.out ("\t--scaling <mode>\t Scaling mode for wallpaper. Can be stretch, fit, fill, default. Must be used before wallpaper provided.\n\
```
Help entries for `--scaling` and `--clamping`

Pasalc commented on src/WallpaperEngine/Assets/ITexture.h:
```diff
@@ -69,6 +69,7 @@ namespace WallpaperEngine::Assets
             NoInterpolation = 1,
             ClampUVs = 2,
             IsGif = 4,
+            ClampUVsBorder = 8,
```
Option for `--clamping`

Pasalc commented on src/WallpaperEngine/Application/CWallpaperApplication.cpp:
```diff
@@ -322,13 +322,13 @@ namespace WallpaperEngine::Application
             context->setWallpaper (
                 it.first,
-                WallpaperEngine::Render::CWallpaper::fromWallpaper (it.second->getWallpaper (), *context, *audioContext)
+                WallpaperEngine::Render::CWallpaper::fromWallpaper (it.second->getWallpaper (), *context, *audioContext, this->m_context.settings.general.screenScalings[it.first])
```
Use `scalling` chosen for this wallpaper

Pasalc commented on src/WallpaperEngine/Application/CWallpaperApplication.cpp:
```diff
@@ -322,13 +322,13 @@ namespace WallpaperEngine::Application
         if (this->m_defaultBackground != nullptr)
             context->setDefaultWallpaper (WallpaperEngine::Render::CWallpaper::fromWallpaper (
-                this->m_defaultBackground->getWallpaper (), *context, *audioContext
+                this->m_defaultBackground->getWallpaper (), *context, *audioContext, this->m_context.settings.render.window.scalingMode
```
Use last chosen `--scalling` option

Pasalc reviewed (commented):
Made simple reviewe, hopefully it will be useful

Pasalc commented on src/WallpaperEngine/Render/CWallpaper.cpp:
```diff
@@ -201,81 +202,25 @@ void CWallpaper::setupShaders ()
-{
-    glBindBuffer (GL_ARRAY_BUFFER, this->m_texCoordBuffer);
-    glBufferData (GL_ARRAY_BUFFER, size, texCoords, GL_STATIC_DRAW);
-}
```
It wasn't used anywhere, so I removed it. Probably should have let it stay.

Almamu commented:
The PR looks pretty good. There's slight adjustements I'd make to how scaling is calculated (mainly using templates to provide the different implementations), but I don't know when I'll have time to properly look at these, so accepting the PR as is. Thanks for your contribution and sorry for the delay in reviewing this!
